proof-systems/src/darlin/tests/final_darlin.rs

```diff
@@ -0,0 +1,353 @@
+//! and wires it to the outside via system inputs.
+use algebra::{AffineCurve, ToConstraintField, UniformRand};
+use r1cs_core::{ConstraintSynthesizer, ConstraintSystem, SynthesisError};
+use crate::darlin::{
+    pcd::{
+        PCD, PCDParameters, PCDCircuit,
+        final_darlin::FinalDarlinPCD,
+        error::PCDError,
+    },
+    accumulators::ItemAccumulator,
+    data_structures::FinalDarlinDeferredData,
+    FinalDarlinProverKey, FinalDarlinVerifierKey, FinalDarlin,
+};
+use poly_commit::{
+    ipa_pc::{InnerProductArgPC, CommitterKey, UniversalParams},
+    Error as PCError
+};
+use rand::{ Rng, RngCore };
+use digest::Digest;
+use r1cs_std::{
+    alloc::AllocGadget,
+    fields::fp::FpGadget,
+    eq::EqGadget,
+};
+use std::ops::MulAssign;
+
+// Dummy Acc used for testing
+pub struct TestAcc {}
+
+impl ItemAccumulator for TestAcc {
+    type AccumulatorProverKey = ();
+    type AccumulatorVerifierKey = ();
+    type AccumulationProof = ();
+    type Item = ();
+
+    fn check_items<R: RngCore>(
+        _vk: &Self::AccumulatorVerifierKey,
+        _accumulators: &[Self::Item],
+        _rng: &mut R
+    ) -> Result<bool, PCError>
+    {
+        Ok(true)
+    }
+
+    fn accumulate_items(
+        _ck: &Self::AccumulatorProverKey,
+        _accumulators: Vec<Self::Item>
+    ) -> Result<(Self::Item, Self::AccumulationProof), PCError>
+    {
+        Ok(((), ()))
+    }
+
+    fn verify_accumulated_items<R: RngCore>(
+        _current_accumulator: &Self::Item,
+        _vk: &Self::AccumulatorVerifierKey,
+        _previous_accumulators: Vec<Self::Item>,
+        _proof: &Self::AccumulationProof,
+        _rng: &mut R
+    ) -> Result<bool, PCError>
+    {
+        Ok(true)
+    }
+}
+
+// Test PCDVk
+pub struct TestPCDVk {}
+
+impl AsRef<()> for TestPCDVk
+{
+    fn as_ref(&self) -> &() { &() }
+}
+
+/// For test purpose only: The previous PCD is just a valid instance of FinalDarlinDeferredData.
+pub struct TestPrevPCD<G1: AffineCurve, G2: AffineCurve>(FinalDarlinDeferredData<G1, G2>);
+
+impl<G1, G2> PCD for TestPrevPCD<G1, G2>
+    where
+        G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+        G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+{
+    type PCDAccumulator = TestAcc;
+    type PCDVerifierKey = TestPCDVk;
+
+    // there is nothing to succinctly verify
+    fn succinct_verify(&self, _vk: &Self::PCDVerifierKey) -> Result<<Self::PCDAccumulator as ItemAccumulator>::Item, PCDError> {
+        Ok(())
+    }
+}
+
+/// The parameters for our test circuit
+#[derive(Clone)]
+pub struct CircuitInfo<G1: AffineCurve, G2: AffineCurve> {
+    pub num_constraints: usize,
+    pub num_variables:   usize,
+    /// used to deduce the number of field elements (well, not consistently)
+    pub dummy_deferred:  FinalDarlinDeferredData<G1, G2>,
+}
```

Comment:
Why a separate ``CircuitInfo``, if its data is just put into the struc ``TestCircuit``?

Reply:
I don't have a constructor for TestCircuit other than init() and init_self() functions. This functions shall have everything that is needed to create an empty/concrete instantiation of the circuit. That's also why it occured to me that in some instances this could not have been possible and I added the possibility to pass some SetupData that should cover these cases